pkg/controller/migration/convert/network.go

```diff
@@ -98,22 +108,158 @@ func handleCalicoCNI(c *components, install *Installation) error {
 		install.Spec.CalicoNetwork.HostPorts = &hp
 	}
 
-	// CNI IPAM pools
-	if len(c.calicoCNIConfig.IPAM.IPv4Pools) != 0 {
-		return ErrIncompatibleCluster{"specifying ipv4_pools in cni config is not supported."}
-	}
-
 	// Other CNI features
 	if c.calicoCNIConfig.FeatureControl.FloatingIPs {
 		return ErrIncompatibleCluster{"floating IPs not supported"}
 	}
 	if c.calicoCNIConfig.FeatureControl.IPAddrsNoIpam {
 		return ErrIncompatibleCluster{"IpAddrsNoIpam not supported"}
 	}
+	if c.calicoCNIConfig.ContainerSettings.AllowIPForwarding {
+		return ErrIncompatibleCluster{"AllowIPForwarding not supported"}
+	}
+
+	return nil
+}
+
+func subhandleCalicoIPAM(c *components, install *Installation) error {
+	if c.calicoCNIConfig.IPAM.Type != "calico-ipam" {
+		// I don't know if this should be an error or not, the way it is written
+		// this check is just double checking what the caller should have already
+		// validated.
+		return nil
+	}
+
+	// TODO: Is there anything we need to do with the Name field on the CNI plugin or IPAM
+	// plugin? We obviously don't have a way currently to persist them.
+
+	// CALICO_NETWORKING_BACKEND
+	netBackend, err := c.node.getEnv(ctx, c.client, containerCalicoNode, "CALICO_NETWORKING_BACKEND")
```

Comment:
I think this PR you're working on is the first scenario where it makes sense for us to have the processing in these subhandlers depend on processing done in the calling handler (handleNetwork). 

have handleNetwork process `CALICO_NETWORKING_BACKEND` by setting fields in the Installation resource. Then, have these handlers use conditionals off of that instead of having to run through that processing a second time.

Alternatively, we move those into functions which can be called from both places.

Reply:
This field doesn't directly translate into a single field so I don't think we can read it and put it into the Installation resource.
I can move it into one function that both subhandle functions call.

Reply:
The way you've moved it into a function looks a million times better. But I just want to continue the original thought -

it seems like it mostly translates into `install.Spec.CalicoNetwork.BGP`. We could have 

```go
	netBackend, err := c.node.getEnv(ctx, c.client, containerCalicoNode, "CALICO_NETWORKING_BACKEND")
	if err != nil {
		return "", err
	}

	switch {
	case netBackend == nil, netBackend == "", *netBackend == "bird":
		install.Spec.CalicoNetwork.BGP = operatorv1.BGPOptionPtr(operatorv1.BGPEnabled)
	case *netBackend == "vxlan", *netBackend == "none":
		install.Spec.CalicoNetwork.BGP = operatorv1.BGPOptionPtr(operatorv1.BGPDisabled)
	default:
		return "", fmt.Errorf("CALICO_NETWORKING_BACKEND %s is not valid", *netBackend)
	}
```

Then the subhandlers could check `install.Spec.CalicoNetwork.BGP`. And if, given the cni / ipam, it's incompatible, they could throw an error.

Reply:
I think that would leave us open to trying to migrate something we don't support. Specifically with HostLocal `vxlan` backend is not valid yet with your suggestion it would look the same as backend of `none`.
There is also a less likely chance with Calico CNI with `none` as the backend.